### PR TITLE
Terraform pagerduty support

### PR DESCRIPTION
**Pull request by benwarfield-usds**
## Related Issue or Background Info

- CDCGov/prime-devops#2 requires us to have a pagerduty integration exposed to our existing alerts systems

## Changes Proposed

- update the metrics alerts to take multiple action IDs instead of just one (so we can send events to multiple receivers)
- update the `global` package to set up terraform-managed pagerduty integrations to Production and Demo services

## To verify:

* terraform best practices
* is this how we want to use our existing Pagerduty services? Should we add new ones instead? Should we import the existing integrations so we don't have multiple integration webhooks for Azure?
* do we want to add any customization to the data as it flows through these webhooks?

## Additional Information

A follow-up PR will make use of these changes in the seven (7) deployed API environments.

### UI
- N/A Any changes to the UI/UX have been approved by design 
- N/A Any new or edited error messages have been approved by design 

### Testing
- [x] Includes a summary of what a code reviewer should verify

### Changes are Backwards Compatible
- N/A Database changes are submitted as a separate PR
- N/A GraphQL schema changes are backward compatible with older version of the front-end

### Security
- N/A Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [x] Any dependencies introduced have been vetted and discussed

## Cloud
- LOL DevOps team has been notified if PR requires ops support

### CODE REVIEW

**Comment by benwarfield-usds** [comment]
This won't pass on its own—need to merge #1503 so that the consumers of the service are updated to match the changed alerting structure.

File: ops/services/pagerduty/_var.tf
Comment (benwarfield-usds) [inline ops/services/pagerduty/_var.tf:12]:
```suggestion
  description = "The short name (very short!) of the action group we are creating."
```

File: ops/global/main.tf
Comment (jdorothy) [inline ops/global/main.tf:52]: Let's actually spell out Pagerduty here. Otherwise, `terraform state list` is pretty unreadable.

File: ops/services/alerts/app_service_metrics/main.tf
Comment (jdorothy) [inline ops/services/alerts/app_service_metrics/main.tf:2]: Can we rename this to something more verbose?

File: ops/demo/alerts.tf
Comment (jdorothy) [inline ops/demo/alerts.tf:11]: I think we need a better way to do this than a straight list. There are probably many things we'd like to alert on that don't require paging someone, but with this configuration we wouldn't be able to tell which `id` is which.

File: ops/demo/alerts.tf
Comment (jdorothy) [inline ops/demo/alerts.tf:12]: Relatedly, do we, in fact, want to page on every current alert? I don't want to be paged in the middle of the night because demo's memory usage is high. See larger question in main comments.

**Review by jdorothy — CHANGES_REQUESTED** [review]

**Comment by jdorothy** [comment]
Larger question for the group: what environments are we required to have up when? I.e. what's our SLA? Our paged alerts should reflect this. Can we implement everything in this PR as written but add a toggle for the actual Slack alerts until we have this conversation? I don't want to hold up the actual work.

File: ops/demo/alerts.tf
Comment (benwarfield-usds) [reply ops/demo/alerts.tf:12]: I assume we don't want to page on non-prod events, but I was proposing to handle that by having the non-prod PagerDuty service not page people. (Which, in fact, is what it isn't doing right now.) It can just send a passive-aggressive emails instead.

Comment (ronaldheft-usds) [reply ops/demo/alerts.tf:12]: The way I configured the lower environments when I setup the PagerDuty services is they will only page during business hours, but notify to Slack at any time. We can customize the hours there or remove the paging completely.

**Comment by ronaldheft-usds** [comment]
> Larger question for the group: what environments are we required to have up when? I.e. what's our SLA? Our paged alerts should reflect this. Can we implement everything in this PR as written but add a toggle for the actual Slack alerts until we have this conversation? I don't want to hold up the actual work.

When I talked with @timbest-usds, I noted the following environments as external and considered production:
* Prod
* Demo
* Training

All other environments were considered non-prod. Does that sound right @timbest-usds?

File: ops/demo/alerts.tf
Comment (ronaldheft-usds) [reply ops/demo/alerts.tf:12]: @jdorothy What do you think about sending all alerts to PagerDuty, and we handle how it alerts from there instead of in our Terraform? That way Terraform only is has the PagerDuty integration code.

File: ops/services/pagerduty/versions.tf
Comment (ronaldheft-usds) [inline ops/services/pagerduty/versions.tf:6]: Sweet, a PagerDuty provider! I don't think we used that on ReportStream. Good to know.

**Comment by timothybest-usds** [comment]
Production environments:
- Prod: Expected availability 24/7, fix issue immediately
- Training: Expected availability Monday to Friday 9-5pm ET, Fix the issue before working hours. if there is an alert on Friday at 11pm it is acceptable to fix that during the day on Saturday or Sunday  
- Demo: Expected availability Monday to Friday 9-5pm ET (when sparkle has a demo)

Non production environments (Pentest, Test, Stg, dev) should only be dealt with during business hours

File: ops/services/pagerduty/versions.tf
Comment (ronaldheft-usds) [reply ops/services/pagerduty/versions.tf:6]: How is this authenticating? env var `PAGERDUTY_TOKEN`?

**Comment by benwarfield-usds** [comment]
I was actually wondering about training and demo, and put them in the non-prod bin (I actually put `stg` in the prod bin on the grounds that `stg` is how we test production infrastructure before releasing it). I think based on Tim's last comment we need three services, not two (or we need some fancy event rules on the pagerduty side, which seems like strictly unnecessary levels of extra work).

File: ops/demo/alerts.tf
Comment (jdorothy) [reply ops/demo/alerts.tf:12]: @ronaldheft-usds That did not occur to me but I love it. Fits perfectly with the general approach of keeping environments/infrastructure code uniform and externalizing the business logic.

File: ops/services/pagerduty/versions.tf
Comment (benwarfield-usds) [reply ops/services/pagerduty/versions.tf:6]: Yes, the environment variable seemed like the best fit.

File: ops/demo/alerts.tf
Comment (benwarfield-usds) [reply ops/demo/alerts.tf:11]: Based on the discussion below, I think we'll actually end up with a single-item list in the short term (since cutting out our direct slack hook and relying on PD seems saner)—probably we can customize the event payload if we want pagerduty to handle individual alerts differently or we can have high-priority and low-priority services, but I'd prefer to wait until we know what our needs are before trying to do anything more complicated.

**Comment by jdorothy** [comment]
> I was actually wondering about training and demo, and put them in the non-prod bin (I actually put `stg` in the prod bin on the grounds that `stg` is how we test production infrastructure before releasing it). I think based on Tim's last comment we need three services, not two (or we need some fancy event rules on the pagerduty side, which seems like strictly unnecessary levels of extra work).

What about just sending every environment to PagerDuty and just muting the non-prod ones there? That way we don't need to implement environment checks for this.

File: ops/services/alerts/app_service_metrics/main.tf
Comment (benwarfield-usds) [reply ops/services/alerts/app_service_metrics/main.tf:2]: Can do—how would `env_title` grab you? Since I assume by "more verbose" you really mean "more understandable for people who don't have perl functions as part of their working memory." 😁

**Comment by ronaldheft-usds** [comment]
> > I was actually wondering about training and demo, and put them in the non-prod bin (I actually put `stg` in the prod bin on the grounds that `stg` is how we test production infrastructure before releasing it). I think based on Tim's last comment we need three services, not two (or we need some fancy event rules on the pagerduty side, which seems like strictly unnecessary levels of extra work).
> 
> What about just sending every environment to PagerDuty and just muting the non-prod ones there? That way we don't need to implement environment checks for this.

That was my original thinking on why there is a PagerDuty service per environment instead of prod/non-prod. Gives more flexibility on choosing hours of alerting / alerting channels.

File: ops/demo/alerts.tf
Comment (jdorothy) [reply ops/demo/alerts.tf:11]: It's more that we should decide if we ever want Slack alerts without PagerDuty alerts or vice-versa. With a straight list we can't do that because we won't be able to tell which `id` is which.

Comment (ronaldheft-usds) [reply ops/demo/alerts.tf:11]: PagerDuty can send alerts just to Slack. A service can be configured to not send texts, app notifications, etc, and only publish via Slack or whatever medium. That's why I'm suggesting we don't need the Slack integration directly in our Terraform. We can do it through PagerDuty.

File: ops/services/alerts/app_service_metrics/main.tf
Comment (jdorothy) [reply ops/services/alerts/app_service_metrics/main.tf:2]: Sure, that works. I _just now_ realized that "uc" was for "uppercase", so I was very confused earlier.

File: ops/global/main.tf
Comment (benwarfield-usds) [reply ops/global/main.tf:52]: Done and moved the state over.

**Review by ronaldheft-usds — APPROVED** [review]

**Pull request merged**
